Fix profile links not showing for vendor staff

midu-01 opened the pull request:
### All Submissions:
* [x] My code follows the [WordPress coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)
* [x] My code satisfies feature requirements
* [x] My code is tested
* [x] My code passes the PHPCS tests
* [x] My code has proper inline documentation
* [ ] I've included related pull request(s) (optional)
* [ ] I've included developer documentation (optional)
* [x] I've added proper labels to this pull request

<!-- Mark completed items with an [x] -->
<!-- You can erase any parts of this template not applicable to your Pull Request. -->

---

### Changes proposed in this Pull Request:

This PR fixes an issue where **vendor social profile URLs were not visible to vendor staff users** in  
**Vendor Dashboard → Settings → Social Profile**.

Previously, the social profile data was fetched strictly based on the logged-in user ID.  
For staff users, this resulted in fetching staff-level data instead of the parent vendor’s store data, causing empty social profile fields.

This change ensures that:
- When a staff member accesses the Social Profile settings, the **parent vendor ID** is resolved and used
- Social profile URLs saved by the vendor are correctly fetched and displayed for staff users
- Existing vendor behavior remains unchanged

---

### Related Pull Request(s)
* N/A

---

### Closes
* Closes https://github.com/getdokan/dokan-pro/issues/5214#event-21494424531
* Closes https://github.com/getdokan/dokan-pro/issues/5293

---

### How to test the changes in this Pull Request:

1. Log in as a **vendor**
2. Navigate to **Vendor Dashboard → Settings → Social Profile**
3. Add social profile URLs (Facebook, Twitter, Instagram, etc.)
4. Save the settings
5. Log out
6. Log in as a **staff user** associated with the same vendor
7. Navigate to **Settings → Social Profile**
8. Verify that:
   - The previously saved social profile URLs are visible
   - The data matches what the vendor configured
9. Update any social profile URL as staff and save
10. Log back in as vendor and verify the changes persist correctly

---

### Changelog entry

**Fix:** Vendor social profiles not visible for staff users

Fixed an issue where social profile URLs configured by a vendor were not displayed when accessed by vendor staff members. The fix ensures staff users correctly load and manage the parent vendor’s social profile data.

---

### Before Changes

- Staff users saw empty social profile fields
- Social URLs saved by the vendor were not loaded
- REST API requests used the staff user ID instead of vendor store ID

(Screenshots if applicable)

---

### After Changes

- Staff users correctly see vendor social profile URLs
- Social profile data is fetched using the resolved vendor ID
- Vendor and staff experiences are consistent and reliable

(Screenshots if applicable)

---

### Feature Video (optional)

N/A

---

### PR Self Review Checklist:
* [x] Code follows existing style guidelines
* [x] Naming is clear and understandable
* [x] KISS principle followed
* [x] No unnecessary duplication (DRY)
* [x] Readable and maintainable code
* [x] No performance regressions
* [x] No unnecessary complexity
* [x] No grammar or spelling errors

---

### FOR PR REVIEWER ONLY:

> As a reviewer, your feedback should be focused on the idea, not the person. Seek to understand, be respectful, and focus on constructive dialog.  
> As a contributor, I will iterate on this PR based on feedback to ensure the best possible solution.

* [ ] Correct — Does the change fully resolve the issue?
* [ ] Secure — No security risks introduced
* [ ] Readable — Code is easy to understand and maintain
* [ ] Elegant — Solution fits the existing architecture


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Centralized vendor authorization: owner vs. staff resolution with conditional exposure of admin-only fields.
  * Store settings endpoints now reuse shared store authorization and response behavior.

* **Bug Fixes**
  * Clearer 400/403/404 responses for invalid or missing stores.
  * Restricted fields are filtered from updates and views for unauthorized roles; sensitive commission fields hidden when appropriate.

* **Tests**
  * Extensive authorization tests for admin, vendor, vendor‑staff, customer, unauthenticated, and edge cases.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

midu-01 committed:
Fix profile links not showing for vendor staff

midu-01 committed:
Fix Social Profile Links Not Showing for Vendor Staff Members

mrabbani commented on includes/REST/StoreController.php:
```diff
@@ -312,15 +312,23 @@ public function get_stores( $request ) {
      */
     public function get_store( $request ) {
         $store_id = (int) $request['id'];
+        $current_user = get_current_user_id();
```
**We may simplify by following the bellow steps:**

- Check the requested id(`$store_id`) is the seller ID using `dokan_is_user_seller( $store_id, true )`. It will ensure that the requested ID is the vendor ID.
-  Otherwise, Fetch the vendor ID for the current user using `dokan_get_current_user_id()`

**Suggested code:**
```php
if ( ! dokan_is_user_seller( $store_id, true ) ) {
   $store_id = dokan_get_current_user_id();
}
//  The existing code
```

midu-01 replied:
Yes, this can be done using the suggested function. At that time, I did not notice the availability of the dokan_is_user_seller() function. Thank you for pointing this out, bhai. I will update the implementation accordingly.

mrabbani commented on includes/REST/StoreController.php:
```diff
@@ -361,11 +369,22 @@ public function update_store_permissions_check( $request ) {
+            return true;
+        }
+
+        return false;
```
Do we need to change the permission?

midu-01 replied:
Bhai,

Yes, we need to handle the permission here. Staff users should not be able to add or update this profile link unless the appropriate permission is explicitly granted. That is why I implemented this logic specifically for controlling staff update permissions.

Please let me know if you would like any adjustments to this approach.

midu-01 committed:
Fix store update permission using dokan_is_user_seller and resolve staff update issue

mrabbani reviewed (changes requested):
- Check user capability
- Ensure that the user has the proper access to the requested Store.

midu-01 committed:
Refactor store access: add check_vendor_authorizable_permission and reuse it

midu-01 committed:
Fix PHPCS error

midu-01 committed:
Fix PHPCS error

midu-01 committed:
Fix PHPCS error

midu-01 committed:
Fix vendor ID resolution for vendor and staff users

midu-01 committed:
Fix vendor ID resolution for vendor and staff users

midu-01 committed:
Fix vendor ID resolution for vendor and staff users

midu-01 committed:
Fix vendor ID resolution for vendor and staff users

midu-01 committed:
Fix vendor ID resolution for vendor and staff users

midu-01 committed:
Fix vendor ID resolution for vendor and staff users

midu-01 committed:
Fix vendor ID resolution for vendor and staff users

mrabbani reviewed (commented):
## Code Review Summary

Overall, this PR addresses an important issue where vendor staff cannot access social profile links. The approach of using a trait for reusable permission logic is good, and the code structure is clean. However, there are **critical bugs** in the implementation that prevent vendor staff from actually accessing stores, despite the documentation stating they should be able to.

---

## ✅ Positive Aspects

1. **Good use of traits** - The `VendorAuthorizable` trait provides reusable permission logic that can be used across controllers
2. **Clear documentation** - Method docblocks clearly explain the intended behavior
3. **Type hints** - Proper use of PHP type hints improves code safety
4. **Security pattern** - The permission checks follow a good pattern (admin → vendor → staff)

---

## 🔴 Critical Bugs (Must Fix)

### 1. Missing Vendor Staff Check in `check_vendor_authorizable_permission()`

**Location:** `includes/Traits/VendorAuthorizable.php:29-51`

**Issue:** The function documentation states "Vendor staff can access only their assigned vendor store," but the implementation **doesn't actually check for vendor staff**. It only checks if the current user is a seller and if their ID matches the vendor ID, which will always fail for vendor staff.

**Current code:**
```php
if ( dokan_is_user_seller( $current_user_id ) ) {
    return (int) $current_user_id === (int) $vendor_id;
}
return false;
```

**Problem:** `dokan_is_user_seller($current_user_id)` returns `true` for vendor staff (since they have the `dokandar` capability), but the subsequent check `$current_user_id === $vendor_id` will **always fail** for staff members since their user ID is different from the vendor ID.

**Fix required:**
```php
$current_user_id = dokan_get_current_user_id();

// Check if current user is the vendor (exclude staff)
if ( dokan_is_user_seller( $current_user_id, true ) ) { // exclude_staff = true
    return (int) $current_user_id === (int) $vendor_id;
}

// Check if current user is vendor staff with access to this vendor
if ( user_can( $current_user_id, 'vendor_staff' ) ) {
    $staff_vendor_id = (int) get_user_meta( $current_user_id, '_vendor_id', true );
    return $staff_vendor_id > 0 && (int) $staff_vendor_id === (int) $vendor_id;
}

return false;
```

### 2. Missing Requested Store ID Check in `update_store_permissions_check()`

**Location:** `includes/REST/StoreController.php:367-374`

**Issue:** The permission check doesn't verify that the requested store ID matches the current user's vendor ID or their staff's vendor ID.

**Current code:**
```php
return $this->check_vendor_authorizable_permission( $this->get_store_id_for_user() );
```

**Problem:** This only checks if the current user has permission to access their own store, but doesn't verify that the requested store ID (from `$request->get_param('id')`) matches.

**Fix required:**
```php
$requested_id = absint( $request->get_param( 'id' ) );
$user_store_id = $this->get_store_id_for_user();

// For vendor staff, we need to resolve the requested ID to their vendor ID
if ( user_can( get_current_user_id(), 'vendor_staff' ) && $requested_id === get_current_user_id() ) {
    $requested_id = $user_store_id;
}

return $this->check_vendor_authorizable_permission( $requested_id );
```

### 3. Potential Logic Issue in `get_store()` Method

**Location:** `includes/REST/StoreController.php:316`

**Issue:** `get_store_id_for_user()` is called with the requested store ID, but this method is designed to convert a user ID to their vendor ID. If the requested ID is already a vendor ID, this could cause issues.

**Current code:**
```php
$store_id = $this->get_store_id_for_user( (int) $request->get_param( 'id' ) );
```

**Concern:** If the API accepts vendor IDs (store IDs), calling `get_store_id_for_user()` with a vendor ID might incorrectly resolve it if that vendor ID happens to be a user ID that's vendor staff. However, since `dokan_is_user_seller($user_id, true)` excludes staff, if the ID is a vendor, it should return correctly. This might be intentional to allow the API to accept either vendor IDs or user IDs, but the logic could be clearer.

**Recommendation:** Add a comment explaining that the API accepts either a vendor ID or user ID, and if it's a user ID for staff, it resolves to their vendor ID.

---

## ⚠️ Code Quality Issues

### 4. Code Duplication in `update_store()` Method

**Location:** `includes/REST/StoreController.php:389-395`

**Issue:** Manual vendor staff handling duplicates logic that should use `get_store_id_for_user()`.

**Current code:**
```php
if ( user_can( $current_user, 'vendor_staff' ) ) {
    $staff_vendor_id = (int) get_user_meta( $current_user, '_vendor_id', true );
    if ( $staff_vendor_id ) {
        $requested_id = $staff_vendor_id;
    }
}
```

**Recommendation:** Use the trait method instead:
```php
// If current user is staff and requesting their own user ID, resolve to vendor ID
if ( user_can( $current_user, 'vendor_staff' ) && $requested_id === $current_user ) {
    $requested_id = $this->get_store_id_for_user( $current_user );
}
```

---

## 🧪 Testing Recommendations

1. **Test vendor staff can access their vendor's store** via GET endpoint
2. **Test vendor staff can update their vendor's store** via PUT/PATCH endpoint  
3. **Test vendor staff cannot access other vendors' stores**
4. **Test regular vendors cannot access other vendors' stores**
5. **Test admins can access any store**
6. **Test edge cases:** staff without `_vendor_id` meta, invalid vendor IDs, etc.

---

## Summary

The **approach is correct**, but the implementation has **critical bugs** that prevent vendor staff from accessing stores. The main issue is that `check_vendor_authorizable_permission()` doesn't check for vendor staff, even though the documentation says it should. Once these bugs are fixed, the implementation should work as intended.

**Priority fixes:**
1. ✅ **Fix Bug #1** - Add vendor staff check in `check_vendor_authorizable_permission()`
2. ✅ **Fix Bug #2** - Add requested store ID verification in `update_store_permissions_check()`
3. ⚠️ **Refactor Bug #4** - Remove code duplication in `update_store()`

Once these are addressed, this PR should be ready to merge. Great work on identifying the issue and creating a reusable solution!

midu-01 committed:
Fix vendor staff store access and improve authorization logic

- Fix vendor staff authorization in store access checks
- Ensure requested store ID is validated in update permission checks
- Align permission logic with documented behavior
- Refactor staff vendor resolution to use shared trait methods
- Improve method naming for clarity and readability
- Add comments to clarify store/user ID resolution logic

mrabbani committed:
Merge branch 'develop' into fix/profile-links-not-showing-for-vendor-staff-5214

mrabbani committed:
feat: Add authorization tests and refactor store validation to use VendorAuthorizable trait

- Add comprehensive test suite for StoreController authorization (20 tests)
- Refactor validate_store_id to use VendorAuthorizable trait method
- Update can_access_vendor_store to support vendor staff authorization
- Centralize vendor ID resolution logic using VendorUtil
- All tests passing (20 tests, 44 assertions)

mrabbani committed:
fix: improve permission handling and vendor ID resolution

- Fix can_access_vendor_store to allow admins to access invalid vendor IDs for proper 404 error handling
- Simplify update_store_permissions_check to use get_vendor_id_for_user consistently
- Improve update_store to properly resolve vendor IDs for vendor staff
- Enhance docblocks with clearer descriptions and role-based behavior explanations
- Ensure public endpoints use exact vendor IDs without resolution

mrabbani committed:
feat: make get_store endpoint truly public with data filtering

- Make get_store endpoint accessible to all users/guests
- Filter sensitive data (payment, enabled status, admin commission) for unauthorized users
- Respect admin settings for hiding vendor info (email, phone, address)
- Show public data to everyone while protecting sensitive information
- Update authorization check to determine data visibility, not access blocking
- Update all test cases to reflect new public endpoint behavior
- Tests now verify filtered data instead of expecting 403 errors

mrabbani committed:
feat: add field restrictions for vendor staff and refactor StoreSettingController

- Add field restriction logic for vendor staff (email, password)
- Add field restriction logic for vendors (admin commission fields)
- Refactor StoreSettingController to extend StoreController for code reuse
- Extract field restriction logic into reusable methods:
  - get_restricted_fields_for_update() - for store updates
  - get_restricted_fields_for_view() - for store data views
- Add comprehensive test suite for StoreSettingController (20 tests)
- Add is_staff_only() method to VendorAuthorizable trait
- Improve code documentation with PHPDoc blocks
- All tests passing (20 tests, 49 assertions)

mrabbani committed:
Add util method create_vendor_staff

dev-shahed commented:
Issue: The social URL not filled for the role vendor staff.

mrabbani committed:
Fix profile links not showing for vendor staff
